navigation/perfect_memory_hog.h

```diff
@@ -0,0 +1,106 @@
+
+namespace BoBRobotics {
+namespace Navigation {
+//------------------------------------------------------------------------
+// BoBRobotics::Navigation::PerfectMemoryHOG
+//------------------------------------------------------------------------
+//! Perfect memory algorithm using HOG features instead of raw image matching
+template<typename RIDFProcessor = BestMatchingSnapshot>
+class PerfectMemoryHOG : public PerfectMemoryBase<RIDFProcessor>
+{
+public:
+    PerfectMemoryHOG(const cv::Size unwrapRes, const unsigned int scanStep = 1,
+                     const filesystem::path outputPath = "snapshots",
+                     const int HOGPixelsPerCell = 10, int HOGOrientations = 8)
+      : PerfectMemoryBase<RIDFProcessor>(unwrapRes, scanStep, outputPath)
+      , m_HOGDescriptorSize(unwrapRes.width * unwrapRes.height *
+                            HOGOrientations / (HOGPixelsPerCell * HOGPixelsPerCell))
+    {
+        std::cout << "Creating perfect memory for HOG features" << std::endl;
+
+        // Configure HOG features
+        m_HOG.winSize = unwrapRes;
+        m_HOG.blockSize = cv::Size(HOGPixelsPerCell, HOGPixelsPerCell);
+        m_HOG.blockStride = m_HOG.blockSize;
+        m_HOG.cellSize = m_HOG.blockSize;
+        m_HOG.nbins = HOGOrientations;
+    }
+
+    //------------------------------------------------------------------------
+    // Constants
+    //------------------------------------------------------------------------
+    const unsigned int m_HOGDescriptorSize;
+
+    //------------------------------------------------------------------------
+    // Declared virtuals
+    //------------------------------------------------------------------------
+    virtual size_t getNumSnapshots() const override
+    {
+        return m_Snapshots.size();
+    }
+
+    virtual const cv::Mat &getSnapshot(size_t) const override
+    {
+        throw std::runtime_error("When using HOG features, snapshots aren't stored");
+    }
+
+protected:
+    // Add a snapshot to memory and return its index
+    virtual size_t addSnapshot(const cv::Mat &image) override
+    {
+        m_Snapshots.emplace_back(m_HOGDescriptorSize);
+        m_HOG.compute(image, m_Snapshots.back());
+        assert(m_Snapshots.back().size() == m_HOGDescriptorSize);
+
+        // Return index of new snapshot
+        return (m_Snapshots.size() - 1);
+    }
+
+    // Calculate difference between memory and snapshot with index
+    virtual float calcSnapshotDifference(const cv::Mat &image, const cv::Mat &imageMask, size_t snapshot) const override
+    {
+        assert(imageMask.empty());
+
+        // Calculate HOG descriptors of image
+        m_HOG.compute(image, m_ScratchDescriptors);
+        assert(m_ScratchDescriptors.size() == m_HOGDescriptorSize);
+
+        // Calculate square difference between image HOG descriptors and snapshot
+        rtransform(m_Snapshots[snapshot], m_ScratchDescriptors, m_ScratchDescriptors, [](float a, float b) {
```

Comment:
I'm not sure how meaningful it would be, but would it make sense to support the different types of RDF comparers?

Reply:
Oh, you mean so that we could have it templated like the ``PerfectMemory`` class?

I think that would make sense. I don't think there is *much* practical difference between doing an abs diff vs. RMS diff, though it would be good to have the choice.

I also meant to say that I'm using abs diff as the default for ``PerfectMemory`` now instead of RMS diff -- mostly because it is faster (it is only slightly faster on my laptop but *way* faster on my old computer, presumably because the laptop has some shiny new SIMD instructions available -- maybe we should run the benchmark on the Tegra).

Reply:
Interesting! I'd guess one's vectorizing the std::transform/has a fast integer->float conversion instruction and one doens't - I'll give it a go on the TX2 tomorrow - I need to re-run some benchmarks on it anyway cos I keep fiddling with/optimising GeNN as I'm supposed to be writing a paper and thus invalidating my results...

Reply:
Btw, just noticed that in ``PerfectMemoryHOG`` you're summing ``m_ScratchDescriptors``, then returning the square root, but you're not dividing by N anywhere. Is this deliberate?

Reply:
...in the ``calcSnapshotDifference`` function, that is.